README.md
# stac-tiler

A rio-tiler plugin to handler STAC items


[![Packaging status](https://badge.fury.io/py/stac-tiler.svg)](https://badge.fury.io/py/stac-tiler)
[![CircleCI](https://circleci.com/gh/developmentseed/stac-tiler.svg?style=svg)](https://circleci.com/gh/cogeotiff/stac-tiler)
[![codecov](https://codecov.io/gh/developmentseed/stac-tiler/branch/master/graph/badge.svg)](https://codecov.io/gh/cogeotiff/stac-tiler)


## Install

```bash
$ pip install pip -U
$ pip install stac-tiler --pre  # stac-tiler is in pre-release 0.0rc1 version 

# Or using source

$ pip install git+http://github.com/developmentseed/stac-tiler
```

## How To

stac-tiler is based on [rio-tiler-crs](https://github.com/cogeotiff/rio-tiler-crs) and [morecantile](https://github.com/developmentseed/morecantile).

## Docs

<details>

```
class STACReader:
    """
    STAC + Cloud Optimized GeoTIFF Reader.

    Examples
    --------
    with STACReader(stac_path) as stac:
        stac.tile(...)

    my_stac = {
        "type": "Feature",
        "stac_version": "1.0.0",
        ...
    }
    with STACReader(None, item=my_stac) as stac:
        stac.tile(...)

    Attributes
    ----------
    filepath: str
        STAC Item path, URL or S3 URL.
    item: Dict, optional
        STAC Item dict.
    tms: morecantile.TileMatrixSet, optional
        TileMatrixSet to use, default is WebMercatorQuad.
    minzoom: int, optional
        Set minzoom for the tiles.
    minzoom: int, optional
        Set maxzoom for the tiles.
    include_assets: Set, optional
        Only accept some assets.
    exclude_assets: Set, optional
        Exclude some assets.
    include_asset_types: Set, optional
        Only include some assets base on their type
    include_asset_types: Set, optional
        Exclude some assets base on their type

    Properties
    ----------
    bounds: tuple[float]
        STAC bounds in WGS84 crs.
    center: tuple[float, float, int]
        STAC item center + minzoom

    Methods
    -------
    tile(0, 0, 0, assets="B01", expression="B01/B02")
        Read a map tile from the COG.
    part((0,10,0,10), assets="B01", expression="B1/B20", max_size=1024)
        Read part of the COG.
    preview(assets="B01", max_size=1024)
        Read preview of the COG.
    point((10, 10), assets="B01")
        Read a point value from the COG.
    stats(assets="B01", pmin=5, pmax=95)
        Get Raster statistics.
    info(assets="B01")
        Get Assets raster info.
    metadata(assets="B01", pmin=5, pmax=95)
        info + stats

    """
```

</details>


- **STACReader.tile()**: Read map tile from STAC assets

```python
with STACReader("stac.json") as stac:
    tile, mask = stac.tile(1, 2, 3, tilesize=256, assets=["red", "green"])

# With expression
with STACReader("stac.json") as stac:
    tile, mask = cog.tile(1, 2, 3, tilesize=256, expression="red/green")
```

- **STACReader.part()**: Read part of STAC assets

```python
with STACReader("stac.json") as stac:
    data, mask = stac.part((10, 10, 20, 20), assets=["red", "green"])

# Limit output size (default is set to 1024)
with STACReader("stac.json") as stac:
    data, mask = stac.part((10, 10, 20, 20), max_size=2000, assets=["red", "green"])

# Read high resolution
with STACReader("stac.json") as stac:
    data, mask = stac.part((10, 10, 20, 20), max_size=None, assets=["red", "green"])

# With expression
with STACReader("stac.json") as stac:
    data, mask = stac.part((10, 10, 20, 20), expression="red/green")
```

- **STACReader.preview()**: Read a preview of STAC assets

```python
with STACReader("stac.json") as stac:
    data, mask = stac.preview(assets=["red", "green"])

# With expression
with STACReader("stac.json") as stac:
    data, mask = stac.preview(expression="red/green")
```

- **STACReader.point()**: Read point value of STAC assets

```python
with STACReader("stac.json") as stac:
    pts = stac.point(-100, 25, assets=["red", "green"])


# With expression
with STACReader("stac.json") as stac:
    pts = stac.point(-100, 25, expression="red/green")
```

- **STACReader.info()**: Return simple metadata for STAC assets

```python
with STACReader("stac.json") as stac:
    info = stac.info("B01")
{
    "B01": {
        "bounds": [23.10607624352815, 31.50517374437416, 24.296464503939944, 32.51933487169619],
        "center": [23.701270373734047, 32.012254308035175, 8],
        "minzoom": 8,
        "maxzoom": 11,
        "band_metadata": [[1, {}]],
        "band_descriptions": [[1, "band1"]],
        "dtype": "uint16",
        "colorinterp": ["gray"],
        "nodata_type": "Nodata"
    }
}
```

- **STACReader.stats()**: Return statistics for STAC assets (Min/Max/Stdev)

```python
with STACReader("stac.json") as stac:
    print(stac.stats(["B01"]))
{
    "B01": {
        "1": {
            "pc": [
                324,
                5046
            ],
            "min": 133,
            "max": 8582,
            "std": 1230.6977195618235,
            "histogram": [
                [
                    199042, 178438, 188457, 118369, 57544, 20622, 9275, 2885, 761, 146
                ],
                [
                    133, 977.9, 1822.8, 2667.7, 3512.6, 4357.5, 5202.4, 6047.3, 6892.2, 7737.099999999999, 8582
                ]
            ]
        }
    }
}
```

- **STACReader.metadata()**: Return info and statistics for STAC assets 

```python
with STACReader("stac.json") as stac:
    print(stac.metadata(["B01"], pmin=5, pmax=95))
{
    "B01": {
        "bounds": [23.10607624352815, 31.50517374437416, 24.296464503939944, 32.51933487169619],
        "center": [23.701270373734047, 32.012254308035175, 8],
        "minzoom": 8,
        "maxzoom": 11,
        "band_metadata": [[1, {}]],
        "band_descriptions": [[1, "band1"]],
        "dtype": "uint16",
        "colorinterp": ["gray"],
        "nodata_type": "Nodata"
        "statistics": {
            "1": {
                "pc": [
                    324,
                    5046
                ],
                "min": 133,
                "max": 8582,
                "std": 1230.6977195618235,
                "histogram": [
                    [
                        199042, 178438, 188457, 118369, 57544, 20622, 9275, 2885, 761, 146
                    ],
                    [
                        133, 977.9, 1822.8, 2667.7, 3512.6, 4357.5, 5202.4, 6047.3, 6892.2, 7737.099999999999, 8582
                    ]
                ]
            }
        }
    }
```


## Contribution & Development

Issues and pull requests are more than welcome.

**dev install**

```bash
$ git clone https://github.com/developmentseed/stac-tiler.git
$ cd stac-tiler
$ pip install -e .[dev]
```

**Python >=3.7 only**

This repo is set to use `pre-commit` to run *isort*, *flake8*, *pydocstring*, *black* ("uncompromising Python code formatter") and mypy when committing new code.

```
$ pre-commit install

$ git add .

$ git commit -m'my change'
isort....................................................................Passed
black....................................................................Passed
Flake8...................................................................Passed
Verifying PEP257 Compliance..............................................Passed
mypy.....................................................................Passed

$ git push origin
```

CHANGES.txt
0.0pre2 (2020-06-05)
------------------
- add `metadata` method

0.0pre1 (2020-06-02)
------------------
- Initial pre-release


tests/test_reader.py
"""Tests for stac_reader."""

import json
import os
from unittest.mock import patch

import morecantile
import pytest
import rasterio
from rasterio.warp import transform_bounds
from stac_tiler import STACReader

from rio_tiler import constants
from rio_tiler.errors import InvalidBandName
from rio_tiler_crs import COGReader

prefix = os.path.join(os.path.dirname(__file__), "fixtures")
STAC_PATH = os.path.join(prefix, "item.json")
ALL_ASSETS = [
    "thumbnail",
    "overview",
    "info",
    "metadata",
    "visual",
    "B01",
    "B02",
    "B03",
    "B04",
    "B05",
    "B06",
    "B07",
    "B08",
    "B8A",
    "B09",
    "B11",
    "B12",
    "AOT",
    "WVP",
    "SCL",
]


class mock_COGReader(COGReader):
    """Mock COGReader."""

    def __enter__(self):
        """Support using with Context Managers."""
        self.filepath
        assert self.filepath.startswith("https://somewhereovertherainbow.io/")
        cog_path = os.path.join(prefix, os.path.basename(self.filepath))
        self.dataset = rasterio.open(cog_path)
        self.bounds = transform_bounds(
            self.dataset.crs, constants.WGS84_CRS, *self.dataset.bounds, densify_pts=21
        )
        return self


@patch("stac_tiler.reader.s3_get_object")
@patch("stac_tiler.reader.requests")
def test_fetch_stac(requests, s3_get):
    with STACReader(STAC_PATH, include_asset_types=None) as stac:
        assert stac.minzoom == 0
        assert stac.maxzoom == 24
        assert stac.tms.identifier == "WebMercatorQuad"
        assert stac.filepath == STAC_PATH
        assert stac.assets == ALL_ASSETS
    requests.assert_not_called()
    s3_get.assert_not_called()

    with STACReader(STAC_PATH) as stac:
        assert stac.minzoom == 0
        assert stac.maxzoom == 24
        assert stac.tms.identifier == "WebMercatorQuad"
        assert stac.filepath == STAC_PATH
        assert "metadata" not in stac.assets
        assert "thumbnail" not in stac.assets
        assert "info" not in stac.assets
    requests.assert_not_called()
    s3_get.assert_not_called()

    with STACReader(STAC_PATH, include_assets={"B01", "B02"}) as stac:
        assert stac.assets == ["B01", "B02"]
    requests.assert_not_called()
    s3_get.assert_not_called()

    with STACReader(STAC_PATH, include_assets={"B01", "B02"}) as stac:
        assert stac.assets == ["B01", "B02"]
    requests.assert_not_called()
    s3_get.assert_not_called()

    with STACReader(
        STAC_PATH, exclude_assets={"overview", "visual", "AOT", "WVP", "SCL"}
    ) as stac:
        assert stac.assets == [
            "B01",
            "B02",
            "B03",
            "B04",
            "B05",
            "B06",
            "B07",
            "B08",
            "B8A",
            "B09",
            "B11",
            "B12",
        ]
    requests.assert_not_called()
    s3_get.assert_not_called()

    with STACReader(STAC_PATH, include_asset_types={"application/xml"}) as stac:
        assert stac.assets == ["metadata"]
    requests.assert_not_called()
    s3_get.assert_not_called()

    with STACReader(
        STAC_PATH,
        include_asset_types={"application/xml", "image/png"},
        include_assets={"metadata", "overview"},
    ) as stac:
        assert stac.assets == ["metadata"]
    requests.assert_not_called()
    s3_get.assert_not_called()

    # Should raise an error in future versions
    with STACReader(STAC_PATH, include_assets={"B1"}) as stac:
        assert not stac.assets
    requests.assert_not_called()
    s3_get.assert_not_called()

    # HTTP
    class MockResponse:
        def __init__(self, data):
            self.data = data

        def json(self):
            return json.loads(self.data)

    with open(STAC_PATH, "r") as f:
        requests.get.return_value = MockResponse(f.read())

    with STACReader(
        "http://somewhereovertherainbow.io/mystac.json", include_assets={"B01"}
    ) as stac:
        assert stac.assets == ["B01"]
    requests.get.assert_called_once()
    s3_get.assert_not_called()
    requests.mock_reset()

    # S3
    with open(STAC_PATH, "r") as f:
        s3_get.return_value = f.read()

    with STACReader(
        "s3://somewhereovertherainbow.io/mystac.json", include_assets={"B01"}
    ) as stac:
        assert stac.assets == ["B01"]
    requests.assert_not_called()
    s3_get.assert_called_once()


@patch("stac_tiler.reader.COGReader", mock_COGReader)
def test_reader_tiles():
    """Test STACReader.tile."""
    tile = morecantile.Tile(z=9, x=289, y=207)

    with STACReader(STAC_PATH) as stac:
        with pytest.raises(InvalidBandName):
            stac.tile(*tile, assets="B1")

        with pytest.raises(Exception):
            stac.tile(*tile)

        data, mask = stac.tile(*tile, assets="B01")
    assert data.shape == (1, 256, 256)
    assert mask.shape == (256, 256)

    with STACReader(STAC_PATH) as stac:
        data, mask = stac.tile(*tile, expression="B01/B02")
    assert data.shape == (1, 256, 256)
    assert mask.shape == (256, 256)

    with STACReader(STAC_PATH) as stac:
        data, mask = stac.tile(*tile, assets=["B01", "B02"])
    assert data.shape == (2, 256, 256)
    assert mask.shape == (256, 256)

    # This is possible but user should not to it ;-)
    # We are reading B01 and B02 and telling rasterio to return twice bidx 1.
    with STACReader(STAC_PATH) as stac:
        data, mask = stac.tile(*tile, assets=["B01", "B02"], indexes=(1, 1))
    assert data.shape == (4, 256, 256)
    assert mask.shape == (256, 256)

    # Power User might use expression for each assets
    with STACReader(STAC_PATH) as stac:
        data, mask = stac.tile(*tile, assets=["B01", "B02"], asset_expression="b1/2")
    assert data.shape == (2, 256, 256)
    assert mask.shape == (256, 256)

    with STACReader(STAC_PATH, tms=morecantile.tms.get("WorldCRS84Quad")) as stac:
        data, mask = stac.tile(4, 1, 2, assets="B01")
    assert data.shape == (1, 256, 256)
    assert mask.shape == (256, 256)


@patch("stac_tiler.reader.COGReader", mock_COGReader)
def test_reader_part():
    """Test STACReader.part."""
    bbox = (23.7, 31.506, 24.1, 32.514)

    with STACReader(STAC_PATH) as stac:
        with pytest.raises(InvalidBandName):
            stac.part(bbox, assets="B1")

        with pytest.raises(Exception):
            stac.part(bbox)

        data, mask = stac.part(bbox, assets="B01")
    assert data.shape == (1, 189, 68)
    assert mask.shape == (189, 68)

    with STACReader(STAC_PATH) as stac:
        data, mask = stac.part(bbox, expression="B04/B02")
    assert data.shape == (1, 1024, 371)
    assert mask.shape == (1024, 371)

    with STACReader(STAC_PATH) as stac:
        data, mask = stac.part(bbox, assets=["B04", "B02"])
    assert data.shape == (2, 1024, 371)
    assert mask.shape == (1024, 371)

    # This is possible but user should not to it ;-)
    # We are reading B01 and B02 and telling rasterio to return twice bidx 1.
    with STACReader(STAC_PATH) as stac:
        data, mask = stac.part(bbox, assets=["B04", "B02"], indexes=(1, 1))
    assert data.shape == (4, 1024, 371)
    assert mask.shape == (1024, 371)

    # Power User might use expression for each assets
    with STACReader(STAC_PATH) as stac:
        data, mask = stac.part(bbox, assets=["B04", "B02"], asset_expression="b1/2")
    assert data.shape == (2, 1024, 371)
    assert mask.shape == (1024, 371)

    with STACReader(STAC_PATH) as stac:
        data, mask = stac.part(bbox, assets="B04", max_size=None)
    assert data.shape == (1, 1129, 408)
    assert mask.shape == (1129, 408)


@patch("stac_tiler.reader.COGReader", mock_COGReader)
def test_reader_preview():
    """Test STACReader.preview."""
    with STACReader(STAC_PATH) as stac:
        with pytest.raises(InvalidBandName):
            stac.preview(assets="B1")

        with pytest.raises(Exception):
            stac.preview()

        data, mask = stac.preview(assets="B01")
    assert data.shape == (1, 183, 183)
    assert mask.shape == (183, 183)

    with STACReader(STAC_PATH) as stac:
        data, mask = stac.preview(expression="B04/B02")
    assert data.shape == (1, 1024, 1024)
    assert mask.shape == (1024, 1024)

    with STACReader(STAC_PATH) as stac:
        data, mask = stac.preview(assets=["B04", "B02"])
    assert data.shape == (2, 1024, 1024)
    assert mask.shape == (1024, 1024)

    # This is possible but user should not to it ;-)
    # We are reading B01 and B02 and telling rasterio to return twice bidx 1.
    with STACReader(STAC_PATH) as stac:
        data, mask = stac.preview(assets=["B04", "B02"], indexes=(1, 1))
    assert data.shape == (4, 1024, 1024)
    assert mask.shape == (1024, 1024)

    # Power User might use expression for each assets
    with STACReader(STAC_PATH) as stac:
        data, mask = stac.preview(assets=["B04", "B02"], asset_expression="b1/2")
    assert data.shape == (2, 1024, 1024)
    assert mask.shape == (1024, 1024)

    with STACReader(STAC_PATH) as stac:
        data, mask = stac.preview(assets="B04", max_size=512)
    assert data.shape == (1, 512, 512)
    assert mask.shape == (512, 512)


@patch("stac_tiler.reader.COGReader", mock_COGReader)
def test_reader_point():
    """Test STACReader.point."""
    lat = 32
    lon = 23.7

    with STACReader(STAC_PATH) as stac:
        with pytest.raises(InvalidBandName):
            stac.point(lon, lat, assets="B1")

        with pytest.raises(Exception):
            stac.point(lon, lat)

        data = stac.point(lon, lat, assets="B01")
    assert len(data) == 1

    with STACReader(STAC_PATH) as stac:
        data = stac.point(lon, lat, expression="B04/B02")
    assert len(data) == 1

    with STACReader(STAC_PATH) as stac:
        data = stac.point(lon, lat, assets=["B04", "B02"])
    assert len(data) == 2

    # This is possible but user should not to it ;-)
    # We are reading B01 and B02 and telling rasterio to return twice bidx 1.
    with STACReader(STAC_PATH) as stac:
        data = stac.point(lon, lat, assets=["B04", "B02"], indexes=(1, 1))
    assert len(data) == 2
    assert len(data[0]) == 2

    # Power User might use expression for each assets
    with STACReader(STAC_PATH) as stac:
        data = stac.point(lon, lat, assets=["B04", "B02"], asset_expression="b1/2")
    assert len(data) == 2


@patch("stac_tiler.reader.COGReader", mock_COGReader)
def test_reader_stats():
    """Test STACReader.stats."""
    with STACReader(STAC_PATH) as stac:
        with pytest.raises(InvalidBandName):
            stac.stats(assets="B1")

        data = stac.stats(assets="B01")
    assert len(data.keys()) == 1
    assert data["B01"]

    with STACReader(STAC_PATH) as stac:
        data = stac.stats(assets=["B04", "B02"])
    assert len(data.keys()) == 2
    assert data["B02"]
    assert data["B04"]


@patch("stac_tiler.reader.COGReader", mock_COGReader)
def test_reader_info():
    """Test STACReader.info."""
    with STACReader(STAC_PATH) as stac:
        with pytest.raises(InvalidBandName):
            stac.info(assets="B1")

        data = stac.info(assets="B01")
    assert len(data.keys()) == 1
    assert data["B01"]

    with STACReader(STAC_PATH) as stac:
        data = stac.info(assets=["B04", "B02"])
    assert len(data.keys()) == 2
    assert data["B02"]
    assert data["B04"]


@patch("stac_tiler.reader.COGReader", mock_COGReader)
def test_reader_metadata():
    """Test STACReader.metadata."""
    with STACReader(STAC_PATH) as stac:
        with pytest.raises(InvalidBandName):
            stac.metadata(assets="B1")

        data = stac.metadata(assets="B01")
    assert len(data.keys()) == 1
    assert data["B01"]
    assert data["B01"]["statistics"]

    with STACReader(STAC_PATH) as stac:
        data = stac.metadata(assets=["B04", "B02"])
    assert len(data.keys()) == 2
    assert data["B02"]
    assert data["B04"]


stac_tiler/reader.py
"""stac_tiler.reader."""

import functools
import json
import multiprocessing
import os
import re
from concurrent import futures
from dataclasses import dataclass, field
from typing import Any, Dict, Iterator, List, Optional, Sequence, Set, Tuple, Union
from urllib.parse import urlparse

import morecantile
import numexpr
import numpy
import requests

from rio_tiler.errors import InvalidBandName
from rio_tiler_crs import COGReader

from .utils import s3_get_object

TMS = morecantile.tms.get("WebMercatorQuad")
MAX_THREADS = int(os.environ.get("MAX_THREADS", multiprocessing.cpu_count() * 5))
DEFAULT_VALID_TYPE = {
    "image/tiff; application=geotiff",
    "image/tiff; application=geotiff; profile=cloud-optimized",
    "image/vnd.stac.geotiff; cloud-optimized=true",
    "image/tiff",
    "image/x.geotiff",
    "image/jp2",
    "application/x-hdf5",
    "application/x-hdf",
}


def _apply_expression(
    blocks: Sequence[str], bands: Sequence[str], data: numpy.ndarray
) -> numpy.ndarray:
    """Apply rio-tiler expression."""
    data = dict(zip(bands, data))
    return numpy.array(
        [
            numpy.nan_to_num(numexpr.evaluate(bloc.strip(), local_dict=data))
            for bloc in blocks
        ]
    )


@functools.lru_cache(maxsize=512)
def fetch(filepath: str) -> Dict:
    """Fetch items."""
    parsed = urlparse(filepath)
    if parsed.scheme == "s3":
        bucket = parsed.netloc
        key = parsed.path.strip("/")
        return json.loads(s3_get_object(bucket, key))

    elif parsed.scheme in ["https", "http", "ftp"]:
        return requests.get(filepath).json()

    else:
        with open(filepath, "r") as f:
            return json.load(f)


def _get_assets(
    item: Dict,
    include: Optional[Set[str]] = None,
    exclude: Optional[Set[str]] = None,
    include_asset_types: Optional[Set[str]] = None,
    exclude_asset_types: Optional[Set[str]] = None,
) -> Iterator:
    """Get Asset list."""
    for asset, asset_info in item["assets"].items():
        _type = asset_info["type"]

        if exclude and asset in exclude:
            continue

        if (exclude_asset_types and _type in exclude_asset_types) or (
            include and asset not in include
        ):
            continue

        if (include_asset_types and _type not in include_asset_types) or (
            include and asset not in include
        ):
            continue

        yield asset


@dataclass
class STACReader:
    """
    STAC + Cloud Optimized GeoTIFF Reader.

    Examples
    --------
    with STACReader(stac_path) as stac:
        stac.tile(...)

    my_stac = {
        "type": "Feature",
        "stac_version": "1.0.0",
        ...
    }
    with STACReader(None, item=my_stac) as stac:
        stac.tile(...)

    Attributes
    ----------
    filepath: str
        STAC Item path, URL or S3 URL.
    item: Dict, optional
        STAC Item dict.
    tms: morecantile.TileMatrixSet, optional
        TileMatrixSet to use, default is WebMercatorQuad.
    minzoom: int, optional
        Set minzoom for the tiles.
    minzoom: int, optional
        Set maxzoom for the tiles.
    include_assets: Set, optional
        Only accept some assets.
    exclude_assets: Set, optional
        Exclude some assets.
    include_asset_types: Set, optional
        Only include some assets base on their type
    include_asset_types: Set, optional
        Exclude some assets base on their type

    Properties
    ----------
    bounds: tuple[float]
        STAC bounds in WGS84 crs.
    center: tuple[float, float, int]
        STAC item center + minzoom

    Methods
    -------
    tile(0, 0, 0, assets="B01", expression="B01/B02")
        Read a map tile from the COG.
    part((0,10,0,10), assets="B01", expression="B1/B20", max_size=1024)
        Read part of the COG.
    preview(assets="B01", max_size=1024)
        Read preview of the COG.
    point((10, 10), assets="B01")
        Read a point value from the COG.
    stats(assets="B01", pmin=5, pmax=95)
        Get Raster statistics.
    info(assets="B01")
        Get Assets raster info.
    metadata(assets="B01", pmin=5, pmax=95)
        info + stats

    """

    filepath: str
    item: Optional[Dict] = None
    tms: morecantile.TileMatrixSet = TMS
    minzoom: int = TMS.minzoom
    maxzoom: int = TMS.maxzoom
    include_assets: Optional[Set[str]] = None
    exclude_assets: Optional[Set[str]] = None
    include_asset_types: Set[str] = field(default_factory=lambda: DEFAULT_VALID_TYPE)
    exclude_asset_types: Optional[Set[str]] = None

    def __enter__(self):
        """Support using with Context Managers."""
        self.item = self.item or fetch(self.filepath)

        # Get Zooms from proj: ?
        self.bounds: Tuple[float, float, float, float] = self.item["bbox"]

        self.assets = list(
            _get_assets(
                self.item,
                include=self.include_assets,
                exclude=self.exclude_assets,
                include_asset_types=self.include_asset_types,
                exclude_asset_types=self.exclude_asset_types,
            )
        )

        return self

    def __exit__(self, *args):
        """Support using with Context Managers."""
        pass

    def _get_href(self, assets: Sequence[str]) -> Sequence[str]:
        """Validate asset names and return asset's url."""
        for asset in assets:
            if asset not in self.assets:
                raise InvalidBandName(f"{asset} is not a valid asset name.")

        return [self.item["assets"][asset]["href"] for asset in assets]

    def _parse_expression(self, expression: str) -> Sequence[str]:
        """Parse rio-tiler band math expression."""
        _re = re.compile("|".join(sorted(self.assets, reverse=True)))
        assets = list(set(re.findall(_re, expression)))
        return assets

    @property
    def center(self) -> Tuple[float, float, int]:
        """Return COG center + minzoom."""
        return (
            (self.bounds[0] + self.bounds[2]) / 2,
            (self.bounds[1] + self.bounds[3]) / 2,
            self.minzoom,
        )

    def _tile(
        self, assets: Sequence[str], *args: Any, **kwargs: Any
    ) -> Tuple[numpy.ndarray, numpy.ndarray]:
        """Assemble multiple rio_tiler.reader.tile."""

        def worker(asset: str):
            with COGReader(asset, tms=self.tms) as cog:
                return cog.tile(*args, **kwargs)

        with futures.ThreadPoolExecutor(max_workers=MAX_THREADS) as executor:
            data, masks = zip(*list(executor.map(worker, assets)))
            data = numpy.concatenate(data)
            mask = numpy.all(masks, axis=0).astype(numpy.uint8) * 255
            return data, mask

    def tile(
        self,
        tile_x: int,
        tile_y: int,
        tile_z: int,
        tilesize: int = 256,
        assets: Union[Sequence[str], str] = None,
        expression: Optional[str] = "",  # Expression based on asset names
        asset_expression: Optional[
            str
        ] = "",  # Expression for each asset based on index names
        **kwargs: Any,
    ) -> Tuple[numpy.ndarray, numpy.ndarray]:
        """Read a TMS map tile from COGs."""
        if isinstance(assets, str):
            assets = (assets,)

        if expression:
            assets = self._parse_expression(expression)

        if not assets:
            raise Exception(
                "assets must be passed either via expression or assets options."
            )

        asset_urls = self._get_href(assets)
        data, mask = self._tile(
            asset_urls, tile_x, tile_y, tile_z, expression=asset_expression, **kwargs,
        )

        if expression:
            blocks = expression.split(",")
            data = _apply_expression(blocks, assets, data)

        return data, mask

    def _part(
        self, assets: Sequence[str], *args: Any, **kwargs: Any
    ) -> Tuple[numpy.ndarray, numpy.ndarray]:
        """Assemble multiple COGReader.part."""

        def worker(asset: str):
            with COGReader(asset) as cog:
                return cog.part(*args, **kwargs)

        with futures.ThreadPoolExecutor(max_workers=MAX_THREADS) as executor:
            data, masks = zip(*list(executor.map(worker, assets)))
            data = numpy.concatenate(data)
            mask = numpy.all(masks, axis=0).astype(numpy.uint8) * 255
            return data, mask

    def part(
        self,
        bbox: Tuple[float, float, float, float],
        max_size: int = 1024,
        assets: Union[Sequence[str], str] = None,
        expression: Optional[str] = "",  # Expression based on asset names
        asset_expression: Optional[
            str
        ] = "",  # Expression for each asset based on index names
        **kwargs: Any,
    ) -> Tuple[numpy.ndarray, numpy.ndarray]:
        """Read part of COGs."""
        if isinstance(assets, str):
            assets = (assets,)

        if expression:
            assets = self._parse_expression(expression)

        if not assets:
            raise Exception(
                "assets must be passed either via expression or assets options."
            )

        asset_urls = self._get_href(assets)
        data, mask = self._part(
            asset_urls, bbox, max_size=max_size, expression=asset_expression, **kwargs
        )

        if expression:
            blocks = expression.split(",")
            data = _apply_expression(blocks, assets, data)

        return data, mask

    def _preview(
        self, assets: Sequence[str], *args: Any, **kwargs: Any
    ) -> Tuple[numpy.ndarray, numpy.ndarray]:
        """Assemble multiple COGReader.preview."""

        def worker(asset: str):
            with COGReader(asset) as cog:
                return cog.preview(*args, **kwargs)

        with futures.ThreadPoolExecutor(max_workers=MAX_THREADS) as executor:
            data, masks = zip(*list(executor.map(worker, assets)))
            data = numpy.concatenate(data)
            mask = numpy.all(masks, axis=0).astype(numpy.uint8) * 255
            return data, mask

    def preview(
        self,
        assets: Union[Sequence[str], str] = None,
        expression: Optional[str] = "",  # Expression based on asset names
        asset_expression: Optional[
            str
        ] = "",  # Expression for each asset based on index names
        **kwargs: Any,
    ) -> Tuple[numpy.ndarray, numpy.ndarray]:
        """Return a preview of COGs."""
        if isinstance(assets, str):
            assets = (assets,)

        if expression:
            assets = self._parse_expression(expression)

        if not assets:
            raise Exception(
                "assets must be passed either via expression or assets options."
            )

        asset_urls = self._get_href(assets)
        data, mask = self._preview(asset_urls, expression=asset_expression, **kwargs)

        if expression:
            blocks = expression.split(",")
            data = _apply_expression(blocks, assets, data)

        return data, mask

    def _point(self, assets: Sequence[str], *args: Any, **kwargs: Any) -> List:
        """Assemble multiple COGReader.point."""

        def worker(asset: str) -> List:
            with COGReader(asset) as cog:
                return cog.point(*args, **kwargs)

        with futures.ThreadPoolExecutor(max_workers=MAX_THREADS) as executor:
            return list(executor.map(worker, assets))

    def point(
        self,
        lon: float,
        lat: float,
        assets: Union[Sequence[str], str] = None,
        expression: Optional[str] = "",  # Expression based on asset names
        asset_expression: Optional[
            str
        ] = "",  # Expression for each asset based on index names
        **kwargs: Any,
    ) -> List:
        """Read a value from COGs."""
        if isinstance(assets, str):
            assets = (assets,)

        if expression:
            assets = self._parse_expression(expression)

        if not assets:
            raise Exception(
                "assets must be passed either via expression or assets options."
            )

        asset_urls = self._get_href(assets)
        point = self._point(asset_urls, lon, lat, expression=asset_expression, **kwargs)

        if expression:
            blocks = expression.split(",")
            point = _apply_expression(blocks, assets, point).tolist()

        return point

    def _stats(self, assets: Sequence[str], *args: Any, **kwargs: Any) -> List:
        """Assemble multiple COGReader.stats."""

        def worker(asset: str) -> Dict:
            with COGReader(asset) as cog:
                return cog.stats(*args, **kwargs)

        with futures.ThreadPoolExecutor(max_workers=MAX_THREADS) as executor:
            return list(executor.map(worker, assets))

    def stats(
        self,
        assets: Union[Sequence[str], str],
        pmin: float = 2.0,
        pmax: float = 98.0,
        **kwargs: Any,
    ) -> Dict:
        """Return array statistics from COGs."""
        if isinstance(assets, str):
            assets = (assets,)

        asset_urls = self._get_href(assets)

        stats = self._stats(asset_urls, pmin, pmax, **kwargs)
        return {asset: stats[ix] for ix, asset in enumerate(assets)}

    def _info(self, assets: Sequence[str]) -> List:
        """Assemble multiple COGReader.stats."""

        def worker(asset: str) -> Dict:
            with COGReader(asset, tms=self.tms) as cog:
                return cog.info

        with futures.ThreadPoolExecutor(max_workers=MAX_THREADS) as executor:
            return list(executor.map(worker, assets))

    def info(self, assets: Union[Sequence[str], str]) -> Dict:
        """Return info from COGs."""
        if isinstance(assets, str):
            assets = (assets,)

        asset_urls = self._get_href(assets)

        infos = self._info(asset_urls)
        return {asset: infos[ix] for ix, asset in enumerate(assets)}

    def _metadata(self, assets: Sequence[str], *args: Any, **kwargs: Any) -> List:
        """Assemble multiple COGReader.stats."""

        def worker(asset: str) -> Dict:
            with COGReader(asset) as cog:
                return cog.metadata(*args, **kwargs)

        with futures.ThreadPoolExecutor(max_workers=MAX_THREADS) as executor:
            return list(executor.map(worker, assets))

    def metadata(
        self,
        assets: Union[Sequence[str], str],
        pmin: float = 2.0,
        pmax: float = 98.0,
        **kwargs: Any,
    ) -> Dict:
        """Return array statistics from COGs."""
        if isinstance(assets, str):
            assets = (assets,)

        asset_urls = self._get_href(assets)

        stats = self._metadata(asset_urls, pmin, pmax, **kwargs)
        return {asset: stats[ix] for ix, asset in enumerate(assets)}
